src/main/java/org/folio/invoices/utils/HelperUtils.java

```diff
@@ -231,4 +242,83 @@ private static boolean isFieldNotEmpty(Object newObject, Object existedObject, S
   private static boolean isFieldChanged(Object newObject, Object existedObject, String field) throws IllegalAccessException {
     return !EqualsBuilder.reflectionEquals(FieldUtils.readDeclaredField(newObject, field, true), FieldUtils.readDeclaredField(existedObject, field, true), true, existedObject.getClass(), true);
   }
+
+  /**
+   * Transform list of id's to CQL query using 'or' operation
+   * @param ids list of id's
+   * @return String representing CQL query to get records by id's
+   */
+  public static String convertIdsToCqlQuery(List<String> ids) {
+    return StreamEx.of(ids).map(id -> "id==" + id).joining(" or ");
+  }
+
+  /**
+   * Wait for all requests completion and collect all resulting objects. In case any failed, complete resulting future with the exception
+   * @param futures list of futures and each produces resulting object on completion
+   * @param <T> resulting type
+   * @return resulting objects
+   */
+  public static <T> CompletableFuture<List<T>> collectResultsOnSuccess(List<CompletableFuture<T>> futures) {
+    return VertxCompletableFuture.allOf(futures.toArray(new CompletableFuture[0]))
+      .thenApply(v -> futures
+        .stream()
+        // The CompletableFuture::join can be safely used because the `allOf` guaranties success at this step
+        .map(CompletableFuture::join)
+        .filter(Objects::nonNull)
+        .collect(toList())
+      );
+  }
+
+  public static double calculateVoucherAmount(Voucher voucher, List<VoucherLine> voucherLines) {
+
+    CurrencyUnit currency = Monetary.getCurrency(voucher.getSystemCurrency());
+
+    MonetaryAmount amount = voucherLines.stream()
+      .map(line -> Money.of(line.getAmount(), currency))
+      .collect(MonetaryFunctions.summarizingMonetary(currency))
+      .getSum();
+
+    return convertToDouble(amount);
+  }
+
+  public static Double calculateVoucherLineAmount(List<FundDistribution> fundDistributions, List<InvoiceLine> invoiceLines, Voucher voucher) {
+    CurrencyUnit invoiceCurrency = Monetary.getCurrency(voucher.getInvoiceCurrency());
+    CurrencyUnit systemCurrency = Monetary.getCurrency(voucher.getSystemCurrency());
+    MonetaryAmount voucherLineAmount = Money.of(0, invoiceCurrency);
+
+    for (FundDistribution fundDistribution : fundDistributions) {
+      InvoiceLine invoiceLine = findLineById(invoiceLines, fundDistribution.getInvoiceLineId());
+      MonetaryAmount subTotal = Money.of(invoiceLine.getTotal(), invoiceCurrency);
+
+      voucherLineAmount = voucherLineAmount.add(subTotal.with(MonetaryOperators.percent(fundDistribution.getPercentage())));
+    }
+
+    MonetaryAmount convertedAmount = voucherLineAmount
```

Comment:
Per my earlier comment(s) this could be simplified to `voucherLine.with(conversion)` if we had the `CurrencyConversion` object...

Reply:
I initially intended to use CurrencyConversion, but then I realized that voucherLineAmount.with(conversion) is a blocking operation and so I've chosen the hard way.

Reply:
I don't think it is... The conversion is obtained from the provider earlier - this is probably the more expensive operation which could block.  Apply the conversion itself should be fairly fast - it's probably doing essentially the same thing you're doing here...

Reply:
Yeah, you're right. Simplified to `voucherLineAmount.with(conversion)`.